app/src/test/groovy/pl/futurecollars/invoicing/controller/InvoiceControllerTest.groovy

```diff
@@ -0,0 +1,146 @@
+import static pl.futurecollars.invoicing.TestHelpers.invoice
+
+@AutoConfigureMockMvc
+@SpringBootTest
+class InvoiceControllerTest extends Specification{
+
+    @Autowired
+    private MockMvc mockMvc
+    @Autowired
+    private JsonService jsonService
+    @Autowired
+    private InMemoryDataBase inMemoryDataBase
+
+    private Invoice originalInvoice = invoice(1)
+
+    def 'should return empty list when database is empty'() {
+        given:
+        def url = "/invoices"
+        expect:
+        def result = mockMvc.perform(get(url))
+                .andExpect(status().isOk())
+                .andReturn()
+                .response
+                .contentAsString
+
+        and:
+        result == "[]"
+    }
+
+    def 'should not return  list when wrong data is sent'() {
+        given:
+        def url = "/invoic2es"
+        expect:
+        def result = mockMvc.perform(get(url))
+                .andExpect(status().isNotFound())
+    }
+
+    def "should return all invoices"() {
+        given:
+        def numberOfInvoices = 3
+        (1..numberOfInvoices).collect { id ->
+            def invoice = invoice(id)
+            invoice.id = inMemoryDataBase.save(invoice)
+        }
+        when:
+        def result = mockMvc.perform(get("/invoices"))
+                .andExpect(status().isOk())
+                .andReturn()
+                .response
+                .contentAsString
+
+        then:
+        def resultList = jsonService.toObject(result, Invoice[])
+        resultList.size() == numberOfInvoices
+
+    }
+
+    def "should save invoice"() {
+        given:
+        Invoice invoice = invoice(1)
+        def url = "/invoice"
+
+        def invoiceJson = jsonService.toJson(invoice)
+        when:
+        def result = mockMvc.perform(post(url).content(invoiceJson)
+                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn()
+                .response
+                .contentAsString
+
+        then:
+        result == "1"
+        def invoiceOptional = inMemoryDataBase.getById(1)
+        invoiceOptional.isPresent()
+    }
+
+    def "should not save invoice when wrong data is sent"() {
+        given:
+        def url = "/invoice"
+
+        expect:
+        mockMvc.perform(post(url).content("siema")
+                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isBadRequest())
+    }
+
+    def "should update invoice"() {
+        given:
+        def updateInvoice = originalInvoice
+        def url = "/invoice/{}"
+
+        def invoiceAsJson = jsonService.toJson(updateInvoice)
+        when:
+        def result = mockMvc.perform(put(url ).content(invoiceAsJson)
+                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isBadRequest())
+        then:
+        updateInvoice.date == originalInvoice.date
+    }
```

Comment:
Ten test jest niepoprawny, nie masz nic w bazie, więc nie da się zrobić update. Nazwa testu wskazuje, że ma to być pozytywna ścieżka, więc sprawdzany status to powinno być ok albo inny z serii 200, a nie bad request.